cl/sentinel/handlers/blocks.go

```diff
@@ -14,24 +14,147 @@
-// 	return ssz_snappy.EncodeAndWrite(stream, &emptyString{}, ResourceUnavaiablePrefix)
-// }
+const MAX_REQUEST_BLOCKS = 96
+
+func (c *ConsensusHandlers) beaconBlocksByRangeHandler(s network.Stream) error {
+	peerId := s.Conn().RemotePeer().String()
+	if err := c.checkRateLimit(peerId, "beaconBlocksByRange", rateLimits.beaconBlocksByRangeLimit); err != nil {
+		ssz_snappy.EncodeAndWrite(s, &emptyString{}, RateLimitedPrefix)
+		return err
+	}
+
+	req := &cltypes.BeaconBlocksByRangeRequest{}
+	if err := ssz_snappy.DecodeAndReadNoForkDigest(s, req, clparams.Phase0Version); err != nil {
+		return err
+	}
+
+	if req.Step != 1 {
+		return errors.New("step must be 1")
+	}
+
+	tx, err := c.indiciesDB.BeginRo(c.ctx)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	// Limit the number of blocks to the count specified in the request.
+	if int(req.Count) > MAX_REQUEST_BLOCKS {
+		req.Count = MAX_REQUEST_BLOCKS
+	}
+
+	beaconBlockRooots, slots, err := beacon_indicies.ReadBeaconBlockRootsInSlotRange(c.ctx, tx, req.StartSlot, req.Count)
+	if err != nil {
+		return err
+	}
+
+	// Read the fork digest
+	forkDigest, err := fork.ComputeForkDigestForVersion(
+		utils.Uint32ToBytes4(c.beaconConfig.GenesisForkVersion),
+		c.genesisConfig.GenesisValidatorRoot,
+	)
+	if err != nil {
+		return err
+	}
+
+	for i, slot := range slots {
+		r, err := c.beaconDB.BlockReader(c.ctx, slot, beaconBlockRooots[i])
+		if err != nil {
+			return err
+		}
+		defer r.Close()
 
-func (c *ConsensusHandlers) blocksByRangeHandler(stream network.Stream) error {
-	log.Trace("Got block by range handler call")
-	return ssz_snappy.EncodeAndWrite(stream, &emptyString{}, ResourceUnavaiablePrefix)
+		if _, err := s.Write(forkDigest[:]); err != nil {
+			return err
+		}
+		_, err = io.Copy(s, r)
+		if err != nil {
+			return err
+		}
+	}
+	return ssz_snappy.EncodeAndWrite(s, &emptyString{}, ResourceUnavaiablePrefix)
 }
 
-func (c *ConsensusHandlers) beaconBlocksByRootHandler(stream network.Stream) error {
-	log.Trace("Got beacon block by root handler call")
-	return ssz_snappy.EncodeAndWrite(stream, &emptyString{}, ResourceUnavaiablePrefix)
+func (c *ConsensusHandlers) beaconBlocksByRootHandler(s network.Stream) error {
+	peerId := s.Conn().RemotePeer().String()
+	if err := c.checkRateLimit(peerId, "beaconBlocksByRoot", rateLimits.beaconBlocksByRootLimit); err != nil {
+		ssz_snappy.EncodeAndWrite(s, &emptyString{}, RateLimitedPrefix)
+		return err
+	}
+
+	var req solid.HashListSSZ = solid.NewHashList(100)
+	if err := ssz_snappy.DecodeAndReadNoForkDigest(s, req, clparams.Phase0Version); err != nil {
+		return err
+	}
+
+	blockRoots := []libcommon.Hash{}
+	for i := 0; i < req.Length(); i++ {
+		blockRoot := req.Get(i)
+		blockRoots = append(blockRoots, blockRoot)
+	}
+
+	tx, err := c.indiciesDB.BeginRo(c.ctx)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	// Limit the number of blocks to the count specified in the request.
+	if len(blockRoots) > MAX_REQUEST_BLOCKS {
```

Comment:
move this check up since we know once we append too many that we should limit (maybe just break from loop if len(blockRoots) >= MAX_REQUEST_BLOCKS

Reply:
@elee1766 That is a really good point. Now move check up, and break the loop once it reaches the  `MAX_REQUEST_BLOCKS`.